src/components/FallenHeroes.tsx

```diff
@@ -20,23 +34,27 @@ function FallenHeroes() {
   function addFallenHero(e: React.FormEvent<HTMLFormElement>) {
     e.preventDefault();
     //add hero to the list if it isnt there yet
-    if (!campaign.restOfAliveHeroes.includes(fallenHero)) {
+    if (!names.includes(fallenHero)) {
+      throw Error("Sorry, your hero is an unknown!");
+    }
+
+    if (names.includes(fallenHero) && !isAlive) {
       throw Error("Sorry, your hero is not among the living!");
     }
-    if (campaign.restOfAliveHeroes.includes(fallenHero)) {
-      const updatedFallen = [...campaign.fallenHeroes, fallenHero];
-      const updatedRestOfAlive = campaign.restOfAliveHeroes.filter(
-        (hero) => hero !== fallenHero
-      );
-      setCampaign({
-        ...campaign,
-        fallenHeroes: updatedFallen,
-        restOfAliveHeroes: updatedRestOfAlive,
+
+    if (names.includes(fallenHero) && isAlive) {
+      campaign.allHeroes.map(function (hero) {
+        if (hero.name === fallenHero) {
+          hero.alive = false;
+          setCampaign(campaign);
```

Comment:
The fact that this code works is actually a coincidence. It will break with future changes.

A component re-renders (i.e. the function executes) when state changes. What you expect here is to re-render based on a state (`campaign`) change because you `setCampaign`. However, your change _mutates_ the state, it does not create a new object reference, which does not meet Reacts's conditions for component re-render. Coincidentally, you also `setFallenHeroes`, which reassignes the object in component state (`fallen`) and causes a re-render. This makes it work because the updated `campaign` object is then read, but you can see how fragile it is - try removing the call to `setFallenHeroes` and console.log `campaign`, there will be no change.

A quick way to solve this is to reassign the `campaign` object like this: `setCampaign({...campaign})`. But you can see how messy this is getting. It should be abstracted away with some state management tool such as Redux. I'd suggest just doing the quick fix and transferring to state management tool in another PR.

The rendering logic in React is crucial to understand, I'd recommend you read [this article](https://dev.to/adevnadia/react-re-renders-guide-why-components-re-render-4ml).